layer/filestore.go

```diff
@@ -191,6 +201,23 @@ func (fms *fileMetadataStore) GetCacheID(layer ChainID) (string, error) {
 	return content, nil
 }
 
+func (fms *fileMetadataStore) GetForeignSource(layer ChainID) (*distribution.Descriptor, error) {
+	content, err := ioutil.ReadFile(fms.getLayerFilename(layer, "foreign-source"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var ref *distribution.Descriptor
+	err = json.Unmarshal(content, &ref)
+	if err != nil {
+		return nil, err
+	}
+	return ref, err
```

Comment:
I think it could be confusing for the function to return a `nil` error but also a `nil` result. The general convention is that `err == nil` means there's a valid value being returned.